---
title: "Candidate Regions 450k"
author: "Carlos Ruiz"
date: "3/3/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warnings = FALSE, message = FALSE)
```

# Introduction

Epimutations detection has two main steps: definition of candidate regions and evaluation of outlier significance. Although there are different algorithms to define epimutations regions, they share common features. In general, epimutations should contain at least 3 CpGs and the distance between contiguous CpGs should be smaller 1Kb. In Illumina 450K array, probes are inequally distributed along the genome, limiting the number of regions that can fulfill the requirements to be considered an epimutation.

The goal of this report is to pre-compute a list of regions that are candidates of becoming an epimutation. 

# Candidate epimutations definition

In order to define the candidate epimutations, we relied on the clustering from bumphunter. We defined a toy dataset with all the CpGs from the Illumina 450K array, and where all the CpGs passed bumphunter cut-off. Then, we run bumphunter and selected those regions with at least 3 CpGs. The code for generating this regions can be found in EpiMutations package.

```{r}
library(GenomicRanges)
library(tidyverse)
library(cowplot)
library(qqman)

load("../results/candidateRegions/candidateRegions.450K.withCREs.GRanges.Rdata")
load("../results/candidateRegions/candidateRegions.450K_hg38.withCREs.GRanges.Rdata")

candRegs <- candRegsGR %>%
  as.data.frame() %>%
  tibble()
```

# Epimutations Descriptives

We found `r length(candRegsGR)` candidate epimutations. 

```{r}
length_all <- candRegs %>%
  ggplot(aes(x = L)) +
  geom_histogram() +
  theme_bw() +
  scale_x_continuous(name = "CpG length")

length_short <- length_all + scale_x_continuous(name = "CpG length", limits = c(3, 32))
plot_grid(length_all, length_short)   
```

The distribution of CpGs per candidate region is very skewed to the left: more than half of the regions has 5 CpGs or less, and `r round(mean(candRegs$L <= 15)*100, 2)`% of regions has 15 CpGs or less. Nonetheless, few candidate regions have a very large number of CpGs, with a maximum in `r max(candRegs$L)`. 

```{r}
width_all <- candRegs %>%
  ggplot(aes(x = width)) +
  geom_histogram() +
  theme_bw() +
  scale_x_continuous(name = "Bases length")

width_short <- width_all + scale_x_continuous(name = "Bases length", limits = c(5, 5e3))
plot_grid(width_all, width_short)   
```

The length distribution of candidate regions is also skewed to the left: more than half of the regions has less htan 750 bp, and `r round(mean(candRegs$width <= 2e3)*100, 2)`% of regions has less than 2Kb. Nonetheless, few candidate regions are large, with a maximum length of `r max(candRegs$width)` bp. 

```{r}
candRegs %>%
  mutate(SNP = cluster,
         BP = start,
         CHR = gsub("chr", "", seqnames),
         CHR = as.numeric(CHR),
         P = width) %>%
  manhattan(logp = FALSE)
```

Most of the largest regions are located in the HLA region, a region with a high number of genes and under recombination. 

# Conversion to hg38

We converted the candidate region from hg19 to hg38 coordinates, using NCBI remap. We selected regions that mapped to one region in hg38 with the same length. This yielded a total of `r length(candRegsGR38)`, the `r round(length(candRegsGR38)/length(candRegsGR)*100, 2)`% of total hg19 regions. 

After converting to hg38, we can use these ranges to annotated to ENCODE cREs. Overall, we mapped `r sum(candRegs$CRE != "")` candidate regions to cREs, representing `r round(mean(candRegs$CRE != "")*100, 2)`% of total candidate regions.

# Overlap with literature epimutations

Finally, we compared our candidate regions with the epimutations described in PMID 32937144. 

```{r}
library(readxl)

epi_lit <- read_excel("../data/Epimutations.PMID32937144.xlsx", skip = 2)
epi_litGR <- makeGRangesFromDataFrame(epi_lit, seqnames.field = "Chr, DMR",
                                      start.field = "Start, DMR (hg19)",
                                      end.field = "End, DMR (hg19)")

eqOver <- findOverlaps(epi_litGR -1 , candRegsGR, type = "equal")
withOver <- findOverlaps(epi_litGR -1 , candRegsGR, type = "within")
withOver <- withOver[!withOver %in% eqOver]
allOver <- findOverlaps(epi_litGR -1 , candRegsGR)
```

`r length(epi_litGR)` were described in PMID 32937144. `r length(eqOver)` were equivalent to one candidate region, while the remaining `r length(withOver)` were inside a candidate region.

```{r}
lenComp <- data.frame(Literature = width(epi_litGR[from(allOver)]),
                      Candidate = width(candRegsGR[to(allOver)])) %>%
  mutate(Difference = Candidate - Literature)
ggplot(lenComp, aes(x = Literature, y = Candidate)) +
  geom_point() +
  geom_abline(slope = 1) +
  theme_bw()
```
As expected, the candidate regions are larger than the regions detected in the literature. The biggest differences are found for the largest candidate regions.

```{r}
diff_all <- ggplot(lenComp, aes(x = Difference)) +
  geom_histogram(binwidth = 500) +
  theme_bw() +
  scale_x_continuous(name = "Length candidate - length literature")

diff_short <- diff_all + scale_x_continuous(name = "Length candidate - length literature", limits = c(-3, 15e3))
plot_grid(diff_all, diff_short)   
```

Most of the regions detected in the literature have a size similar to a candidate region: around half of the regions in the literature have a size less than 1Kb smaller than a candidate region, and this proportion reaches `r round(mean(lenComp$Difference <= 3e3)*100, 2)`% for differences smaller than 3Kb.

As a conclussion, candidate regions are good starting points to define epimutations. Based on these results, we can use them for different approaches:

- Speed up computations: focus only on CpGs included in epimutations
- Add cREs annotation
- Homogeneize epimutations naming: this can be very useful when comparing different studies and for reporting
- Develop new algorithms
